backends-clickhouse/src/main/delta-22/org/apache/spark/sql/delta/Snapshot.scala

```diff
@@ -0,0 +1,522 @@
+  }
+
+  protected lazy val checkpointFileIndexOpt: Option[DeltaLogFileIndex] = {
+    assertLogFilesBelongToTable(path, logSegment.checkpoint)
+    DeltaLogFileIndex(DeltaLogFileIndex.CHECKPOINT_FILE_FORMAT, logSegment.checkpoint)
+  }
+
+  def getCheckpointMetadataOpt: Option[CheckpointMetaData] = checkpointMetadataOpt
+
+  def deltaFileSizeInBytes(): Long = deltaFileIndexOpt.map(_.sizeInBytes).getOrElse(0L)
+  def checkpointSizeInBytes(): Long = checkpointFileIndexOpt.map(_.sizeInBytes).getOrElse(0L)
+
+  protected lazy val fileIndices: Seq[DeltaLogFileIndex] = {
+    checkpointFileIndexOpt.toSeq ++ deltaFileIndexOpt.toSeq
+  }
+
+  /**
+   * Loads the file indices into a DataFrame that can be used for LogReplay.
+   *
+   * In addition to the usual nested columns provided by the SingleAction schema, it should provide
+   * two additional columns to simplify the log replay process: [[ACTION_SORT_COL_NAME]] (which,
+   * when sorted in ascending order, will order older actions before newer ones, as required by
+   * [[InMemoryLogReplay]]); and [[ADD_STATS_TO_USE_COL_NAME]] (to handle certain combinations of
+   * config settings for delta.checkpoint.writeStatsAsJson and delta.checkpoint.writeStatsAsStruct).
+   */
+  protected def loadActions: DataFrame = {
+    val dfs = fileIndices.map(index => Dataset.ofRows(spark, deltaLog.indexToRelation(index)))
+    dfs
+      .reduceOption(_.union(_))
+      .getOrElse(emptyDF)
+      .withColumn(ACTION_SORT_COL_NAME, input_file_name())
+      .withColumn(ADD_STATS_TO_USE_COL_NAME, col("add.stats"))
+  }
+
+  protected def emptyDF: DataFrame =
+    spark.createDataFrame(spark.sparkContext.emptyRDD[Row], logSchema)
+
+  override def logInfo(msg: => String): Unit = {
+    super.logInfo(s"[tableId=${deltaLog.tableId}] " + msg)
+  }
+
+  override def logWarning(msg: => String): Unit = {
+    super.logWarning(s"[tableId=${deltaLog.tableId}] " + msg)
+  }
+
+  override def logWarning(msg: => String, throwable: Throwable): Unit = {
+    super.logWarning(s"[tableId=${deltaLog.tableId}] " + msg, throwable)
+  }
+
+  override def logError(msg: => String): Unit = {
+    super.logError(s"[tableId=${deltaLog.tableId}] " + msg)
+  }
+
+  override def logError(msg: => String, throwable: Throwable): Unit = {
+    super.logError(s"[tableId=${deltaLog.tableId}] " + msg, throwable)
+  }
+
+  override def toString: String =
+    s"${getClass.getSimpleName}(path=$path, version=$version, metadata=$metadata, " +
+      s"logSegment=$logSegment, checksumOpt=$checksumOpt)"
+
+  override def filesForScan(filters: Seq[Expression], keepNumRecords: Boolean): DeltaScan = {
+    val deltaScan = super.filesForScan(filters, keepNumRecords)
+
+    DeltaScan.apply(
+      deltaScan.version,
+      deltaScan.files
+        .map(
+          addFile => {
+            val addFileAsKey = AddFileAsKey(addFile)
+            ClickhouseSnapshot.fileStatusCache.get(addFileAsKey)
```

Comment:
this `fileStatusCache` only cache the `AddMergeTreeParts` but not reduce the time for listing from delta log ?  the `deltaScan.files` seems it will call select action.

Reply:
fixed